src/soter/openssl/soter_rsa_key_3.c

```diff
@@ -0,0 +1,644 @@
+{
+    soter_status_t err = SOTER_FAIL;
+    int rsa_mod_size;
+    BIGNUM* rsa_n = NULL;
+    BIGNUM* rsa_e = NULL;
+    EVP_PKEY* pkey = (EVP_PKEY*)(*engine_key);
+    const uint32_t* pub_exp;
+    OSSL_PARAM* params = NULL;
+    OSSL_PARAM_BLD* bld = NULL;
+    EVP_PKEY_CTX* ctx = NULL;
+
+    if (key_length != be32toh(key->size)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    /* Validate tag */
+    if (memcmp(key->tag, RSA_PUB_KEY_PREF, strlen(RSA_PUB_KEY_PREF)) != 0) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (SOTER_SUCCESS != soter_verify_container_checksum(key)) {
+        return SOTER_DATA_CORRUPT;
+    }
+
+    switch (key->tag[3]) {
+    case RSA_SIZE_TAG_1024:
+        rsa_mod_size = 128;
+        break;
+    case RSA_SIZE_TAG_2048:
+        rsa_mod_size = 256;
+        break;
+    case RSA_SIZE_TAG_4096:
+        rsa_mod_size = 512;
+        break;
+    case RSA_SIZE_TAG_8192:
+        rsa_mod_size = 1024;
+        break;
+    default:
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (key_length < rsa_pub_key_size(rsa_mod_size)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    pub_exp = (const uint32_t*)((unsigned char*)(key + 1) + rsa_mod_size);
+    switch (be32toh(*pub_exp)) {
+    case RSA_3:
+    case RSA_F4:
+        break;
+    default:
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    rsa_n = BN_new();
+    rsa_e = BN_new();
+    if (!rsa_n || !rsa_e) {
+        err = SOTER_NO_MEMORY;
+        goto free_exponents;
+    }
+
+    if (!BN_set_word(rsa_e, be32toh(*pub_exp))) {
+        goto free_exponents;
+    }
+
+    if (!BN_bin2bn((const unsigned char*)(key + 1), rsa_mod_size, rsa_n)) {
+        goto free_exponents;
+    }
+
+    bld = OSSL_PARAM_BLD_new();
+    if (bld == NULL) {
+        err = SOTER_NO_MEMORY;
+        goto free_exponents;
+    }
+
+    if (!OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_N, rsa_n)) {
+        err = SOTER_FAIL;
+        goto free_exponents;
+    }
+
+    if (!OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, rsa_e)) {
+        err = SOTER_FAIL;
+        goto free_exponents;
+    }
+
+    params = OSSL_PARAM_BLD_to_param(bld);
+    if (params == NULL) {
+        err = SOTER_NO_MEMORY;
+        goto free_exponents;
+    }
+
+    ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
+    if (ctx == NULL) {
+        err = SOTER_NO_MEMORY;
+        goto free_exponents;
+    }
+
+    if (!EVP_PKEY_fromdata_init(ctx)) {
+        err = SOTER_FAIL;
+        goto free_exponents;
+    }
+
+    if (!EVP_PKEY_fromdata(ctx, (EVP_PKEY**)engine_key, EVP_PKEY_KEYPAIR, params)) {
+        err = SOTER_FAIL;
+        goto free_exponents;
+    }
+
+    err = SOTER_SUCCESS;
+    // goto free_builder; // pass through, BIGINTs were not consumed, need to free everything
+
+free_exponents:
+    BN_free(rsa_n);
+    BN_free(rsa_e);
+free_builder:
+    EVP_PKEY_CTX_free(ctx);
+    OSSL_PARAM_free(params);
+    OSSL_PARAM_BLD_free(bld);
```

Comment:
`free_builder` label is never used.

Since those functions follow free() convention, you can use a single label `out` or `err`.